### PR TITLE
feat: Add raid detection + auto ban

### PR DESCRIPTION
**Pull request by Gerrit0**
Fiddle to experiment with this, since getting a ton of joins on Discord isn't the easiest way to test this... https://jsfiddle.net/Gerrit0/kg20yn8x/

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -20,12 +20,19 @@ const client = new CookiecordClient(
...
+				Intents.FLAGS.GUILD_MEMBERS | // guildMemberAdd
+				Intents.FLAGS.GUILD_MESSAGES | // message, messageDelete, messageUpdate
+				Intents.FLAGS.GUILD_MESSAGE_REACTIONS, // messageReactionAdd, messageReactionRemove
+		},
 		partials: ['REACTION', 'MESSAGE', 'USER', 'CHANNEL'],
 		disableMentions: 'everyone',
 	},
 );
 
+client.setMaxListeners(15);
```
Comment (Gerrit0) [inline src/index.ts:34]: Cookiecord was warning about too many "message" listeners, this made it shut up

**Comment by Gerrit0** [comment]
Note: This requires the "Server Members Intent" permission to be turned on at https://discord.com/developers/applications

File: src/modules/mod.ts [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 import { rulesChannelId } from '../env';
 
 // Most job posts are in this format:
 // > [FOR HIRE][REMOTE][SOMETHING ELSE]
 // > Hi, I'm ponyman6000. Hire me!
 const jobPostRegex = /^(?:\[[A-Z ]+\]){2,}\n/i;
 
+// If more than RAID_JOINS occur in RAID_SECONDS, we're probably getting raided.
+const RAID_JOINS = 4;
+const RAID_SECONDS = 7;
```
Comment (autumnblazey) [inline src/modules/mod.ts:12]: i feel this *might* be a little bit sensitive

Comment (Gerrit0) [reply src/modules/mod.ts:12]: This is what Vortex is currently set to use

Comment (autumnblazey) [reply src/modules/mod.ts:12]: i see 👍

File: src/modules/mod.ts [hunk 3]
```diff
@@ -20,4 +23,30 @@ export class ModModule extends Module {
...
+	async onJoin(member: GuildMember) {
+		const now = Date.now();
+		const lowerBound = now - RAID_SECONDS * 1000;
+
+		this.joins.push(member);
+
+		// Clean up joins from from outside of our raid threshold.
+		while ((this.joins[0].joinedTimestamp ?? 0) < lowerBound) {
+			this.joins.splice(0, 1);
+			this.bannedUpTo--;
```
Comment (d-fischer) [inline src/modules/mod.ts:37]: Probably goes negative a lot while we're not being raided.
```suggestion
			this.bannedUpTo = Math.max(this.bannedUpTo - 1, 0);
```

**Review by d-fischer — CHANGES_REQUESTED** [review]

File: src/modules/mod.ts [hunk 4]
```diff
@@ -20,4 +23,30 @@ export class ModModule extends Module {
...
+		}
+
+		// Ban everyone that just joined
+		if (this.joins.length > RAID_JOINS) {
+			const start = this.bannedUpTo;
+			this.bannedUpTo = this.joins.length;
+
+			await Promise.all(
+				this.joins
+					.slice(start)
```
Comment (d-fischer) [inline src/modules/mod.ts:47]: For the record, the negative number in `bannedUpTo` might lead to things like `.slice(-1)` here, which might mean that we detect raids properly, but don't ban the first `RAID_JOINS - 1` accounts.

File: src/modules/mod.ts (hunk 2 above)
Comment (ronthecookie) [reply src/modules/mod.ts:12]: I feel like maybe this could be relative to normal traffic instead of absolute numbers? Might be a bit overkill though.

Comment (autumnblazey) [reply src/modules/mod.ts:12]: i wonder what the current joins look like? then probably tweak these numbers to fit the needs of the server specifically

**Comment by tjjfvi** [comment]
Ignoring the incorrect negative values, `bannedUpTo` will only ever be `0`, or `this.joins.length - 1`. Feels like this could be a boolean instead.

**Comment by Gerrit0** [comment]
> Ignoring the incorrect negative values, `bannedUpTo` will only ever be `0`, or `this.joins.length - 1`. Feels like it could be a boolean instead.

This is what I started doing originally, but it seemed easier to reason about with the index... consider this:

```
1s - A joins
2s - B joins
3s - C joins
4s - D joins
4s - Raid threshold met, ban A, B, C, D
6s - E joins
6s - joins is [C, D, E]
6s - Raid threshold not met, don't ban E.
6.5s - F joins
6.5s - joins is [C, D, E, F]
6.5s - Raid threshold met, ban E, F
```

Yes, you could figure out how to do this with a boolean rather than an index, but I bet it would be more complicated. The negative problem is fixed now, so I'm pretty sure this works as intended.

**Comment by tjjfvi** [comment]
Ah, I hadn't considered a case like that. In that case, a boolean wouldn't be sufficient, so index is good 👍 .

**Comment by tjjfvi** [comment]
It isn't too difficult to beat this system, even without knowing the `RAID_JOINS`/`RAID_SECONDS` variables ahead of time. https://jsfiddle.net/bygwtL6a/

**Comment by Gerrit0** [comment]
Updated: https://jsfiddle.net/Gerrit0/kg20yn8x/24/

**Comment by ckiee** [comment]
Do you still want to work on this? The jsfiddle makes me think there's a lot of room for false-positives and a PR implementing [Phisherman](https://phisherman.gg/) support would probably be on the safer side.

**Comment by tjjfvi** [comment]
We have Vortex now so this is probably no longer needed

**Comment by ckiee** [comment]
It also seems to be stale. Closing.

**Pull request closed without merging**